<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Contains methods for reading and displaying contents of
the wgs_contig_mean_cov_&amp;lt;phenotype&amp;gt;.csv file output from DRAGEN.
This file contains the estimated coverage for all contigs, and an autosomal
estimated coverage."><title>ContigMeanCovFile R6 Class — ContigMeanCovFile • dracarys</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ContigMeanCovFile R6 Class — ContigMeanCovFile"><meta property="og:description" content="Contains methods for reading and displaying contents of
the wgs_contig_mean_cov_&amp;lt;phenotype&amp;gt;.csv file output from DRAGEN.
This file contains the estimated coverage for all contigs, and an autosomal
estimated coverage."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dracarys</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/umccr/dracarys/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>ContigMeanCovFile R6 Class</h1>
      <small class="dont-index">Source: <a href="https://github.com/umccr/dracarys/blob/HEAD/R/ContigMeanCovFile.R" class="external-link"><code>R/ContigMeanCovFile.R</code></a></small>
      <div class="d-none name"><code>ContigMeanCovFile.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Contains methods for reading and displaying contents of
the <code>wgs_contig_mean_cov_&lt;phenotype&gt;.csv</code> file output from DRAGEN.
This file contains the estimated coverage for all contigs, and an autosomal
estimated coverage.</p>
    </div>


    <div class="section level2">
    <h2 id="super-class">Super class<a class="anchor" aria-label="anchor" href="#super-class"></a></h2>
    <p><code><a href="File.html">dracarys::File</a></code> -&gt; <code>ContigMeanCovFile</code></p>
    </div>
    <div class="section level2">
    <h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a></h2>
    
<div class="section">
<h3 id="public-methods">Public methods<a class="anchor" aria-label="anchor" href="#public-methods"></a></h3>

<ul><li><p><a href="#method-read"><code>ContigMeanCovFile$read()</code></a></p></li>
<li><p><a href="#method-plot"><code>ContigMeanCovFile$plot()</code></a></p></li>
<li><p><a href="#method-clone"><code>ContigMeanCovFile$clone()</code></a></p></li>
</ul></div><p><details open><summary>Inherited methods</summary><ul><li><p><span class="pkg-link" data-pkg="dracarys" data-topic="File" data-id="bname"><a href="File.html#method-bname"><code>dracarys::File$bname()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="dracarys" data-topic="File" data-id="initialize"><a href="File.html#method-initialize"><code>dracarys::File$initialize()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="dracarys" data-topic="File" data-id="print"><a href="File.html#method-print"><code>dracarys::File$print()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="dracarys" data-topic="File" data-id="type"><a href="File.html#method-type"><code>dracarys::File$type()</code></a></span></p></li>
</ul></details></p><hr><a id="method-read"></a><div class="section">
<h3 id="method-read-">Method <code><a href="read.html">read()</a></code><a class="anchor" aria-label="anchor" href="#method-read-"></a></h3>
<p>Reads the <code>wgs_contig_mean_cov_&lt;phenotype&gt;.csv</code> file output from DRAGEN.</p><div class="section">
<h4 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">ContigMeanCovFile</span><span class="op">$</span><span class="fu">read</span><span class="op">(</span>keep_alt <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h4>
<p></p><div class="arguments"><dl><dt><code>keep_alt</code></dt>
<dd><p>Keep the ALT + Mito chromosomes?</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns">Returns<a class="anchor" aria-label="anchor" href="#returns"></a></h4>
<p>tibble with the following columns:</p><ul><li><p>label: file label.</p></li>
<li><p>chrom: contig name.</p></li>
<li><p>n_bases: number of bases aligned to contig (excludes bases from
duplicate marked reads, reads with MAPQ=0, and clipped bases).</p></li>
<li><p>coverage: col2 / contig length</p></li>
</ul></div>

</div><p></p><hr><a id="method-plot"></a><div class="section">
<h3 id="method-plot-">Method <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code><a class="anchor" aria-label="anchor" href="#method-plot-"></a></h3>
<p>Plots the <code>wgs_contig_mean_cov_&lt;phenotype&gt;.csv</code> files.</p><div class="section">
<h4 id="usage-1">Usage<a class="anchor" aria-label="anchor" href="#usage-1"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">ContigMeanCovFile</span><span class="op">$</span><span class="fu">plot</span><span class="op">(</span>top_alt_n <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-1">Arguments<a class="anchor" aria-label="anchor" href="#arguments-1"></a></h4>
<p></p><div class="arguments"><dl><dt><code>top_alt_n</code></dt>
<dd><p>Number of top covered alt contigs to plot per phenotype.</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns-1">Returns<a class="anchor" aria-label="anchor" href="#returns-1"></a></h4>
<p>A ggplot2 object with chromosomes on X axis, and coverage on Y axis.</p>
</div>

</div><p></p><hr><a id="method-clone"></a><div class="section">
<h3 id="method-clone-">Method <code>clone()</code><a class="anchor" aria-label="anchor" href="#method-clone-"></a></h3>
<p>The objects of this class are cloneable with this method.</p><div class="section">
<h4 id="usage-2">Usage<a class="anchor" aria-label="anchor" href="#usage-2"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">ContigMeanCovFile</span><span class="op">$</span><span class="fu">clone</span><span class="op">(</span>deep <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-2">Arguments<a class="anchor" aria-label="anchor" href="#arguments-2"></a></h4>
<p></p><div class="arguments"><dl><dt><code>deep</code></dt>
<dd><p>Whether to make a deep clone.</p></dd>


</dl><p></p></div>
</div>

</div>

    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">x1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/wgs/SEQC-II.wgs_contig_mean_cov_normal.csv.gz"</span>, package <span class="op">=</span> <span class="st">"dracarys"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">x2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/wgs/SEQC-II.wgs_contig_mean_cov_tumor.csv.gz"</span>, package <span class="op">=</span> <span class="st">"dracarys"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">cc1</span> <span class="op">&lt;-</span> <span class="va">ContigMeanCovFile</span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">x1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">cc2</span> <span class="op">&lt;-</span> <span class="va">ContigMeanCovFile</span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">x2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="read.html">read</a></span><span class="op">(</span><span class="va">cc1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 25 × 4</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    label     chrom     n_bases coverage</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> SEQC-II_N chr1  <span style="text-decoration: underline;">11</span>391<span style="text-decoration: underline;">872</span>260     49.4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> SEQC-II_N chr2  <span style="text-decoration: underline;">11</span>910<span style="text-decoration: underline;">292</span>998     49.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> SEQC-II_N chr3   <span style="text-decoration: underline;">9</span>864<span style="text-decoration: underline;">967</span>562     49.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> SEQC-II_N chr4   <span style="text-decoration: underline;">9</span>380<span style="text-decoration: underline;">962</span>565     49.4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> SEQC-II_N chr5   <span style="text-decoration: underline;">8</span>851<span style="text-decoration: underline;">326</span>991     49.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> SEQC-II_N chr6   <span style="text-decoration: underline;">7</span>013<span style="text-decoration: underline;">593</span>668     41.2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> SEQC-II_N chr7   <span style="text-decoration: underline;">7</span>641<span style="text-decoration: underline;">057</span>161     48.1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> SEQC-II_N chr8   <span style="text-decoration: underline;">7</span>051<span style="text-decoration: underline;">795</span>336     48.7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> SEQC-II_N chr9   <span style="text-decoration: underline;">5</span>706<span style="text-decoration: underline;">387</span>709     46.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> SEQC-II_N chr10  <span style="text-decoration: underline;">6</span>468<span style="text-decoration: underline;">177</span>560     48.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># … with 15 more rows</span></span>
<span class="r-in"><span><span class="fu"><a href="read.html">read</a></span><span class="op">(</span><span class="va">cc2</span>, keep_alt <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 3,367 × 4</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    label     chrom     n_bases coverage</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> SEQC-II_T chr1  <span style="text-decoration: underline;">19</span>825<span style="text-decoration: underline;">282</span>617     86.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> SEQC-II_T chr2  <span style="text-decoration: underline;">20</span>173<span style="text-decoration: underline;">660</span>434     83.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> SEQC-II_T chr3  <span style="text-decoration: underline;">17</span>397<span style="text-decoration: underline;">714</span>315     87.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> SEQC-II_T chr4  <span style="text-decoration: underline;">15</span>471<span style="text-decoration: underline;">335</span>389     81.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> SEQC-II_T chr5  <span style="text-decoration: underline;">15</span>376<span style="text-decoration: underline;">687</span>254     85.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> SEQC-II_T chr6  <span style="text-decoration: underline;">13</span>348<span style="text-decoration: underline;">447</span>303     78.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> SEQC-II_T chr7  <span style="text-decoration: underline;">16</span>991<span style="text-decoration: underline;">597</span>908    107. </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> SEQC-II_T chr8  <span style="text-decoration: underline;">13</span>202<span style="text-decoration: underline;">798</span>183     91.2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> SEQC-II_T chr9  <span style="text-decoration: underline;">10</span>085<span style="text-decoration: underline;">223</span>132     82.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> SEQC-II_T chr10 <span style="text-decoration: underline;">10</span>799<span style="text-decoration: underline;">872</span>299     81.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># … with 3,357 more rows</span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cc1</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="ContigMeanCovFile-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer></div>

  

  

  </body></html>

