<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Contains methods for reading and displaying contents of
the vc_metrics.csv file output from DRAGEN, which contains variant calling metrics.
reported for each sample in multi sample VCF and gVCF files.
Based on the run case, metrics are reported either as standard VARIANT
CALLER or JOINT CALLER. Metrics are reported both for the raw
(PREFILTER) and hard filtered (POSTFILTER) VCFs.
PON (Panel of Normals) and COSMIC filtered variants are counted as
though they are PASS variants in the POSTFILTER VCFs metrics,
which may result in higher than expected variant counts in the
POSTFILTER VCF metrics."><title>VCMetricsFile R6 Class — VCMetricsFile • dracarys</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="VCMetricsFile R6 Class — VCMetricsFile"><meta property="og:description" content="Contains methods for reading and displaying contents of
the vc_metrics.csv file output from DRAGEN, which contains variant calling metrics.
reported for each sample in multi sample VCF and gVCF files.
Based on the run case, metrics are reported either as standard VARIANT
CALLER or JOINT CALLER. Metrics are reported both for the raw
(PREFILTER) and hard filtered (POSTFILTER) VCFs.
PON (Panel of Normals) and COSMIC filtered variants are counted as
though they are PASS variants in the POSTFILTER VCFs metrics,
which may result in higher than expected variant counts in the
POSTFILTER VCF metrics."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dracarys</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.10.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">Intro</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/flames.html">FlameStories</a>
    <a class="dropdown-item" href="../articles/developers.html">DevNotes</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/umccr/dracarys/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>VCMetricsFile R6 Class</h1>
      <small class="dont-index">Source: <a href="https://github.com/umccr/dracarys/blob/HEAD/R/dragen.R" class="external-link"><code>R/dragen.R</code></a></small>
      <div class="d-none name"><code>VCMetricsFile.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Contains methods for reading and displaying contents of
the <code>vc_metrics.csv</code> file output from DRAGEN, which contains variant calling metrics.
reported for each sample in multi sample VCF and gVCF files.
Based on the run case, metrics are reported either as standard VARIANT
CALLER or JOINT CALLER. Metrics are reported both for the raw
(PREFILTER) and hard filtered (POSTFILTER) VCFs.
PON (Panel of Normals) and COSMIC filtered variants are counted as
though they are PASS variants in the POSTFILTER VCFs metrics,
which may result in higher than expected variant counts in the
POSTFILTER VCF metrics.</p>
    </div>


    <div class="section level2">
    <h2 id="super-class">Super class<a class="anchor" aria-label="anchor" href="#super-class"></a></h2>
    <p><code><a href="File.html">dracarys::File</a></code> -&gt; <code>VCMetricsFile</code></p>
    </div>
    <div class="section level2">
    <h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a></h2>
    
<div class="section">
<h3 id="public-methods">Public methods<a class="anchor" aria-label="anchor" href="#public-methods"></a></h3>

<ul><li><p><a href="#method-VCMetricsFile-read"><code>VCMetricsFile$read()</code></a></p></li>
<li><p><a href="#method-VCMetricsFile-clone"><code>VCMetricsFile$clone()</code></a></p></li>
</ul></div><p><details open><summary>Inherited methods</summary><ul><li><span class="pkg-link" data-pkg="dracarys" data-topic="File" data-id="bname"><a href="File.html#method-bname"><code>dracarys::File$bname()</code></a></span></li>
<li><span class="pkg-link" data-pkg="dracarys" data-topic="File" data-id="initialize"><a href="File.html#method-initialize"><code>dracarys::File$initialize()</code></a></span></li>
<li><span class="pkg-link" data-pkg="dracarys" data-topic="File" data-id="print"><a href="File.html#method-print"><code>dracarys::File$print()</code></a></span></li>
<li><span class="pkg-link" data-pkg="dracarys" data-topic="File" data-id="type"><a href="File.html#method-type"><code>dracarys::File$type()</code></a></span></li>
</ul></details></p><hr><a id="method-VCMetricsFile-read"></a><div class="section">
<h3 id="method-read-">Method <code><a href="read.html">read()</a></code><a class="anchor" aria-label="anchor" href="#method-read-"></a></h3>
<p>Reads the <code>vc_metrics.csv</code> file output from DRAGEN.</p><div class="section">
<h4 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">VCMetricsFile</span><span class="op">$</span><span class="fu">read</span><span class="op">(</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="returns">Returns<a class="anchor" aria-label="anchor" href="#returns"></a></h4>
<p>tibble with the following columns:</p><ul><li><p>category</p></li>
<li><p>sample</p></li>
<li><p>var: variable name</p></li>
<li><p>count: count value</p></li>
<li><p>pct: percent value</p></li>
</ul></div>

</div><p></p><hr><a id="method-VCMetricsFile-clone"></a><div class="section">
<h3 id="method-clone-">Method <code>clone()</code><a class="anchor" aria-label="anchor" href="#method-clone-"></a></h3>
<p>The objects of this class are cloneable with this method.</p><div class="section">
<h4 id="usage-1">Usage<a class="anchor" aria-label="anchor" href="#usage-1"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">VCMetricsFile</span><span class="op">$</span><span class="fu">clone</span><span class="op">(</span>deep <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h4>
<p></p><div class="arguments"><dl><dt><code>deep</code></dt>
<dd><p>Whether to make a deep clone.</p></dd>


</dl><p></p></div>
</div>

</div>

    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/wgs/SEQC-II.vc_metrics.csv.gz"</span>, package <span class="op">=</span> <span class="st">"dracarys"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">vm</span> <span class="op">&lt;-</span> <span class="va">VCMetricsFile</span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">vm</span><span class="op">$</span><span class="fu">read</span><span class="op">(</span><span class="op">)</span> <span class="co"># or read(vm)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 45 × 5</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    category  sample           var                    count    pct</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> summary   <span style="color: #949494;">""</span>               Number of samples          1  <span style="color: #BB0000;">NA</span>   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> summary   <span style="color: #949494;">""</span>               Reads Processed   <span style="text-decoration: underline;">2</span>617<span style="text-decoration: underline;">162</span>106  <span style="color: #BB0000;">NA</span>   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> summary   <span style="color: #949494;">""</span>               Child Sample              <span style="color: #BB0000;">NA</span>  <span style="color: #BB0000;">NA</span>   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> prefilter <span style="color: #949494;">"</span>SEQC-II_tumour<span style="color: #949494;">"</span> Total                 <span style="text-decoration: underline;">177</span>975 100   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> prefilter <span style="color: #949494;">"</span>SEQC-II_tumour<span style="color: #949494;">"</span> Biallelic             <span style="text-decoration: underline;">177</span>975 100   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> prefilter <span style="color: #949494;">"</span>SEQC-II_tumour<span style="color: #949494;">"</span> Multiallelic               0   0   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> prefilter <span style="color: #949494;">"</span>SEQC-II_tumour<span style="color: #949494;">"</span> SNPs                  <span style="text-decoration: underline;">143</span>829  80.8 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> prefilter <span style="color: #949494;">"</span>SEQC-II_tumour<span style="color: #949494;">"</span> Insertions (Hom)           0   0   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> prefilter <span style="color: #949494;">"</span>SEQC-II_tumour<span style="color: #949494;">"</span> Insertions (Het)       <span style="text-decoration: underline;">16</span>354   9.19</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> prefilter <span style="color: #949494;">"</span>SEQC-II_tumour<span style="color: #949494;">"</span> Deletions (Hom)            0   0   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 35 more rows</span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>, <a href="https://github.com/victorskl" class="external-link">Victor San Kho Lin</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

