---
author: "University of Melbourne Centre for Cancer Research"
date: "`r Sys.time()`"
output:
  html_document:
    toc: true
    theme: cosmo
    css: style.css
    code_download: true
  rmdformats::material:
    highlight: kate
params:
  title: "UMCCR TSO500 ctDNA Summary Report"
description: "UMCCR TSO500 ctDNA Summary Report"
title: "`r params$title`"
---

```{r knitr_opts, include=F}
knitr::opts_chunk$set(
  collapse = TRUE, echo = FALSE,
  warning = FALSE, message = FALSE
)
```

```{r load_pkgs}
{
  require(dplyr)
  require(tidyr)
  require(purrr)
  require(tibble)
  require(readr)
  # require(dracarys) # only needed for session info
  require(glue, include.only = "glue")
  require(here, include.only = "here")
  require(knitr, include.only = "kable")
  require(DT)
  require(fs)
}
```

```{r data_setup, eval=FALSE}
options(scipen = 999) # disable scientific notation
options(width = 150)
pmeta <- here("nogit/data_portal/2023-05-07_workflows_5f43da12-0b6a-41ce-86c9-9ee62df8792e.csv") |>
  dracarys::cttso_metadata()

d <- here("nogit/warehouse/cttso") |>
  dir_ls(recurse = TRUE) |>
  as_tibble_col(column_name = "path") |>
  mutate(sbj = basename(dirname(path))) |>
  # slice(1:10) |>
  rowwise() |>
  mutate(dat = list(read_rds(path))) |>
  select(sbj, dat) |>
  unnest_wider(dat, simplify = FALSE)
d |>
  write_rds(here("nogit/tso/rds/tso_summary_object_d_2022-12-21.rds"))
```

```{r data_setup2}
d <- read_rds(here("nogit/tso/rds/tso_summary_object_d_2022-12-21.rds"))
options(scipen = 999) # disable scientific notation
options(width = 150)
## align_collapse_fusion_caller_metrics
alcolfus <- d |>
  pull("tso__align_collapse_fusion_caller_metrics") |>
  set_names(d$sbj) |>
  bind_rows(.id = "sbj")
alcolfus_hist <- alcolfus |>
  filter(grepl("Hist", .data$name))
alcolfus <- alcolfus |>
  filter(!grepl("Hist", .data$name))

## target_region_coverage
cvg <- d |>
  pull("tso__target_region_coverage") |>
  set_names(d$sbj) |>
  bind_rows(.id = "sbj")

## fragment_length_hist
fraglenhist <- d |>
  pull("tso__fragment_length_hist") |>
  set_names(d$sbj) |>
  bind_rows(.id = "sbj")

## msi
msi <- d |>
  pull("tso__msi") |>
  set_names(d$sbj) |>
  bind_rows(.id = "sbj")

## tmb
tmb <- d |>
  pull("tso__tmb") |>
  set_names(d$sbj) |>
  bind_rows(.id = "sbj")

## combined_variant_output
combinedvars <- d |>
  select("sbj", "tso__combined_variant_output") |>
  unnest_wider("tso__combined_variant_output", simplify = FALSE)

#### Sequencing Run Details
## all NULL..
# d_comb |>
#  pull(`Sequencing Run Details`) |>
#  set_names(d$sbj) |>
#  bind_rows(.id = "sbj")

fus <- d |>
  pull("tso__fusions_csv") |>
  set_names(d$sbj) |>
  bind_rows(.id = "sbj")

## sample_analysis_results
sampres <- d |>
  select("sbj", "tso__sample_analysis_results") |>
  unnest_wider("tso__sample_analysis_results", simplify = FALSE)

trace <- d |>
  pull("tso__tmb_trace_tsv") |>
  set_names(d$sbj) |>
  bind_rows(.id = "sbj")
```

```{r funcs, echo=F}
# can use this to view smallish tables, just don't use it on big ones!
dt_view <- function(x, caption = NULL, scroll_y = 200) {
  x |>
    datatable(
      filter = list(position = "top", clear = FALSE, plain = TRUE),
      class = "cell-border display compact",
      rownames = FALSE,
      extensions = c("Scroller", "Buttons", "KeyTable"),
      options = list(
        scroller = TRUE, scrollY = scroll_y, scrollX = TRUE,
        autoWidth = FALSE, keys = TRUE,
        buttons = c("csv", "excel"), dom = "Blfrtip"
      ),
      caption = caption
    )
}
```

## `SampleAnalysisResults.json.gz` {.tabset #SampleAnalysisResults}

### SNVs

```{r res_snvs}
sampres_vars <- sampres |>
  select(sbj, vars) |>
  unnest_wider(vars)

sampres_vars |>
  select("sbj", "snvs") |>
  unnest("snvs") |>
  glimpse()
```

### CNVs

```{r res_cnvs}
sampres_vars |>
  select("sbj", "cnvs") |>
  unnest("cnvs") |>
  glimpse()
```

### SampleInfo

```{r res_sample_info}
sampres |>
  pull("sample_info") |>
  set_names(sampres$sbj) |>
  bind_rows(.id = "sbj") |>
  glimpse()
```

### SoftwareConfig {.tabset}

#### datasources

```{r res_sw_config}
sampres |>
  hoist("software_config", data_sources = "data_sources") |>
  select("sbj", "data_sources") |>
  unnest("data_sources") |>
  glimpse()
```

#### other

```{r res_sw_config2}
sampres |>
  hoist("software_config", other = "other") |>
  select("sbj", "other") |>
  unnest("other") |>
  glimpse()
```

### Biomarkers (MSI/TMB)

```{r res_biom}
sampres |>
  hoist("biomarkers",
    msi_pct_unstable_sites = list("msi", "msi_pct_unstable_sites"),
    msi_additional_metrics = list("msi", "additional_metrics"),
    tmb_per_mb = list("tmb", "tmb_per_mb"),
    tmb_additional_metrics = list("tmb", "additional_metrics"),
  ) |>
  select(
    "sbj",
    "msi_pct_unstable_sites", "msi_additional_metrics",
    "tmb_per_mb", "tmb_additional_metrics"
  ) |>
  unnest("msi_additional_metrics") |>
  select("sbj", "msi_pct_unstable_sites", msi_subjsd = "value", "tmb_additional_metrics") |>
  unnest("tmb_additional_metrics") |>
  select(-"RAD") |>
  pivot_wider(names_from = c("name"), values_from = "value", names_prefix = "tmb_") |>
  glimpse()
```

### QC1

```{r res_qc1}
sampres |>
  select("sbj", "sample_metrics_qc") |>
  unnest("sample_metrics_qc") |>
  glimpse()
```

### QC2

```{r res_qc2}
sampres |>
  select("sbj", "sample_metrics_expanded") |>
  unnest("sample_metrics_expanded") |>
  glimpse()
```

---

## `Fusions.csv`

```{r fusions}
fus |>
  glimpse()
```


## `tmb.json.gz`

```{r tmb}
tmb |>
  glimpse()
```

## `msi.json.gz`

```{r msi}
msi |>
  glimpse()
```

## `CombinedVariantOutput.tsv` {.tabset #CombinedVariantOutput}

### SNVs

```{r cv_snvs}
combinedvars[["Small Variants"]] |>
  set_names(d$sbj) |>
  bind_rows(.id = "sbj") |>
  glimpse()
```

### CNVs

```{r cv_cnvs}
combinedvars[["Copy Number Variants"]] |>
  set_names(d$sbj) |>
  bind_rows(.id = "sbj") |>
  select(-"...3") |>
  glimpse()
```

### Fusions

```{r cv_fusions}
combinedvars[["DNA Fusions"]] |>
  set_names(d$sbj) |>
  bind_rows(.id = "sbj") |>
  glimpse()
```

### Analysis Details

```{r cv_adet}
combinedvars[["Analysis Details"]] |>
  set_names(d$sbj) |>
  bind_rows(.id = "sbj") |>
  glimpse()
```

### TMB

```{r cv_tmb}
combinedvars[["TMB"]] |>
  set_names(d$sbj) |>
  bind_rows(.id = "sbj") |>
  glimpse()
```

### MSI

```{r cv_msi}
combinedvars[["MSI"]] |>
  set_names(d$sbj) |>
  bind_rows(.id = "sbj") |>
  glimpse()
```

---

## `TMB_Trace.tsv`


```{r tmb_trace}
trace |>
  select(
    chrom = "Chromosome", pos = "Position", ref = "RefCall", alt = "AltCall", "VAF",
    dp = "Depth", "GermlineFilterProxi", "IncludedInTMBNumerator", everything()
  ) |>
  glimpse()
```

## `fragment_length_hist.json.gz`

```{r fraglenhist_plot}
fraglenhist |>
  glimpse()
```

## `TargetRegionCoverage.json.gz`

```{r cvg}
cvg |>
  glimpse()
```

## `AlignCollapseFusionCaller_metrics.json.gz` {.tabset #AlignCollapseFusionCaller_metrics}

### CoverageSummary

```{r met1}
alcolfus |>
  filter(section == "CoverageSummary") |>
  mutate(
    value = ifelse(grepl("Aligned .* in genome", name), glue("{value} ({percent}%)"), value)
  ) |>
  select(-c("section", "percent")) |>
  pivot_wider(names_from = "name", values_from = "value") |>
  glimpse()
```

### MappingAligningPerRg

```{r met2}
alcolfus |>
  filter(section == "MappingAligningPerRg") |>
  mutate(
    value = ifelse(!is.na(percent), glue("{value} ({percent}%)"), value)
  ) |>
  select(-c("section", "percent")) |>
  pivot_wider(names_from = "name", values_from = "value") |>
  glimpse()
```

### MappingAligningSummary

```{r met3}
alcolfus |>
  filter(section == "MappingAligningSummary") |>
  mutate(
    value = ifelse(!is.na(percent), glue("{value} ({percent}%)"), value)
  ) |>
  select(-c("section", "percent")) |>
  filter(!is.na(name)) |> # temp
  pivot_wider(names_from = "name", values_from = "value") |>
  glimpse()
```

### RunTime

```{r met4}
alcolfus |>
  filter(section == "RunTime") |>
  mutate(value = glue("{value} ({percent}%)")) |>
  select(-c("section", "percent")) |>
  pivot_wider(names_from = "name", values_from = "value") |>
  glimpse()
```

### SvSummary

```{r met5}
alcolfus |>
  filter(.data$section == "SvSummary") |>
  mutate(value = as.numeric(.data$value)) |>
  select(-c("section", "percent")) |>
  pivot_wider(names_from = "name", values_from = "value") |>
  glimpse()
```

### TrimmerStatistics

```{r met6}
# for these the percent is NA
numeric_cols <- c(
  "Average bases trimmed per read", "Average bases trimmed per trimmed read",
  "Average input read length", "Total input bases", "Total input bases R1",
  "Total input bases R2", "Total input reads"
)
alcolfus |>
  filter(section == "TrimmerStatistics") |>
  mutate(
    value = ifelse(!is.na(percent), glue("{value} ({percent}%)"), value)
  ) |>
  select(-c("section", "percent")) |>
  pivot_wider(names_from = "name", values_from = "value") |>
  mutate(across(all_of(numeric_cols), as.numeric)) |>
  glimpse()
```

### UmiStatistics

```{r met7}
# for these the percent is NA
numeric_cols <- c(
  "Collapsible region size standard deviation", "Consensus pairs emitted",
  "Max collapsible region size (num reads)", "Mean collapsible region size (num reads)",
  "Mean family depth", "Min collapsible region size (num reads)",
  "Number of collapsible regions", "Number of reads", "On target mean family depth",
  "On target number of reads", "Total number of families"
)

alcolfus |>
  filter(section == "UmiStatistics") |>
  mutate(
    value = ifelse(!is.na(percent), glue("{value} ({percent}%)"), value)
  ) |>
  select(-c("section", "percent")) |>
  pivot_wider(names_from = "name", values_from = "value") |>
  mutate(across(all_of(numeric_cols), as.numeric)) |>
  glimpse()
```

---

## Session Info

```{r session_info, echo=FALSE, eval=FALSE}
pkgs_of_interest <- c("base", "dracarys", "tidyverse", "tidyselect", "ggplot2", "dplyr", "tidyr", "readr")
# need dracarys for session info
si <- dracarys::session_info_kable(pkgs_of_interest)
si$si_pkg
si$si_pl
```
